### PR TITLE
Issue 67

### PR DESCRIPTION
**Pull request by skizzay**
After reviewing the comments on #67, I decided that it's probably best to start off with a mechanism that the user can register their own generators rather than having conan try to discover.  I feel that by starting with this design and when we determine a discovery mechanism, we can simply add code to use the existing design (i.e. the discovery mechanism would registers the generators that it discovers).  Always easier to add than to change/remove.

I broke up conans.client.generators into 2 parts: conans.model.conan_generator and created a module for conans.client.generators.  The idea is that the model is used for extending the Generator base class and using the manager to register new generators.  Of course, if someone doesn't like an existing generator, they can remove it and add their own.  The client.generators module behaves exactly the same.  I just refactored a bit to put each generator into its own file (it was getting pretty big).

I don't have an example or tests at this point.  I wanted to generate discussion and see if this is the design/approach we want to take.  This is nowhere ready to merge into develop.

### CODE REVIEW

**Comment by skizzay** [comment]
I'll also need to grab the two pull requests relating to qmake and qbs generators.  I'll do that once they have been merged into develop.

**Comment by lasote** [comment]
I have merged the generator's PRs. Would be great if you can update the PR.

Thanks!

File: conans/client/installer.py [hunk 1]
```diff
@@ -289,6 +289,7 @@ def _build_package(self, export_folder, src_folder, build_folder, conan_file, ou
             output.info('Copying sources to build folder')
             shutil.copytree(src_folder, build_folder, symlinks=True)
         os.chdir(build_folder)
+        conan_file._conanfile_directory = build_folder
```
Comment (memsharded) [inline conans/client/installer.py]: I'd say this has been already done in another PR, maybe is the conflicting issue. Also, even if just two lines, they are a bit unrelated to this PR, so maybe deserve their own PR.

File: conans/model/conan_generator.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    @property
+    def deps_build_info(self):
+        return self._deps_build_info
+
+    @property
+    def build_info(self):
+        return self._build_info
+
+    @property
+    def content(self):
```
Comment (memsharded) [inline conans/model/conan_generator.py]: Typically I don't feel ABC are necessary, and this approach is enough. In this case, it is fine as-is, but I wouldn't oppose to creating an ABC for the Generator class and abstractproperties for these, they may help for generators without tests providing more detailed error info.

**Comment by memsharded** [comment]
Thanks for the refactor of splitting generators into their own modules, it was becoming very necessary, excellent!

Yes, I like the registration approach, I think is a good approach, clear, and a good base for letting users add their own generators easily.

File: conans/client/installer.py (hunk 1 above)
Comment (skizzay) [inline conans/client/installer.py]: Yeah... So...  This was part of another PR, but I think I must have forgot to push this file for whatever reason.  This is to get conan test to work under certain cases where the implementation of the ConanFile is using the conanfile_directory property.

File: conans/model/conan_generator.py (hunk 2 above)
Comment (skizzay) [inline conans/model/conan_generator.py]: Makes sense.  I'll switch this to be an ABC with abstract properties.
